### DISCUSSION TITLE
Mundane warfare improvements

### DISCUSSION THREAD

**Discussion by Anprivate** [fedorovvl/tso_client#247] [Ideas]
Sorry, long text, so I've written it in native language :)

Новая фича в виде mundane warfare - просто отличная, спасибо. Но, как обычно, хочется с перламутровыми пуговицами.
Попробую немного потеоретизировать. Для атаки нужно знать:
1. Какой генерал используется
2. Какая у него загрузка (сколько и каких войск)
3. Где он должен находиться для атаки
4. Кого он должен атаковать
5. Если генералов несколько, то последовательность их посыла и задержка атаки относительно предыдущего генерала (для блоков)

И логично всё это сохранить в один темплейт. Кроме того, есть генералы которые на данном этапе нужно ставить на нужную точку, но при этом не посылать в атаку. И вроде сейчас всё это есть, но при этом как-то разбросано и несколько нелогично.
Например, почему для задания последовательности и задержек используется отдельная таблица, причём доступная только в момент сохранения темплейта? Не проще ли просто в основную таблицу добавить поле задержки и ввести возможность сортировки генералов по последовательности посыла в атаку?
В темплейт не сохраняется требуемое количество войск, что потенциально может привести к ошибке.
В табличке выбора времени задержки присутствуют все выбранные генералы, а не только те, у которых выбраны цели.

То есть что бы хотелось бы иметь:
1. Табличка должна быть одна и в ней должна быть как последовательность посыла в атаку, так и задержки
2. В темплейт должно сохраняться всё, включая нужную загрузку генералов.
3. В темплейт должны сохраняться в том числе не атакущие генералы (просто для перемещения в нужную позицию с нужной загрузкой).
4. При загрузке темплейта должна проверяться правильность загрузки и быть возможность загрузки (как в Army).

**Comment by Anprivate** [comment]
По результатам использования во время евента - когда всё это нужнее всего. В моём случае практически неюзабельно. Поясняю.
а) во время евента адвенчуры делаются друг за другом. То есть закончился один ywc, сразу запускается другой. У меня схема посыла войск сделана так, что генералы для первых ударов и последующих - разные. И к моменту начала следующего адвенчура часть генералов с предыдущего уже вернулись и готовы к отправке. Следующие отправляются по мере возвращения. Логично было бы сделать заранее темплейт с загрузкой всех нужных генералов и загружать/отправлять тех, которые в наличии. Но нет, если хоть один генерал из темплейта недоступен - я не могу ничего загрузить. То есть либо делать несколько отдельных темплейтов для каждой волны, либо вообще не использовать эту фичу
б) то же самое по позиционированию и отправке в атаку - у меня всё идёт волнами. То есть часть генералов перебазируется на новые точки, но не атакует, часть перебазируется и атакует. Такой алгоритм сейчас вообще нереализуем.

В общем, моё видение, как было бы лучше сделать (отменяющее предыдущее):
а) сделать три отдельных команды - загрузка генералов, отправка генералов на заданные позиции, атака выбранных генералов
б) первые две команды сделать исполнимыми даже если не все генералы, входящие в темплейт, сейчас доступны - загружаем и расставляем тех, которые есть
в) в атаке проверять исходную позицию и загрузку генералов и не начинать атаку без полной готовности
г) пока не знаю как, но очень хорошо было бы реализовать возможность замены генерала на аналогичного. То есть у меня 15 ghost'ов, например. И все абсолютно одинаково заскиллены. То есть чтобы в темплейте писался просто тип генерала и либо автоматически, либо вручную я мог отсутствующего генерала подменить на аналогичного свободного. Но это уже из категории мечт

**Reply by BasilMas** [reply]
> б) то же самое по позиционированию и отправке в атаку - у меня всё идёт волнами...Такой алгоритм сейчас вообще нереализуем.

Все pеализуемо пpи наличии теpпения и вpемени
Именно pазные темплейты пpекpасно спpавляются, напpимеp, в Юном я себе сделал так
![image](https://user-images.githubusercontent.com/76757249/230769919-ece2dc3a-69e3-44c1-8442-bc7e99934312.png)

пока часть отвоевавших возвpащается, свободные пеpеезжают   

+ если генералов хватает - pеимпоpт своего же шаблона отлично и главное быстpо масштабиpует закладку на дpугой пул генералов за минуту

**Comment by skelgaard** [comment]
there is a small x infront on move and after on attack and if the generals is not avaliable, then you can click this X and just send thoose that are ready.
But i can understand what you say and if some smart way for this could be done, it would be nice yes

**Reply by Anprivate** [reply]
Main problem is not with attack, but load and position

**Reply by skelgaard** [reply]
same goes for both load and positioning, can be done all places

**Reply by 00dcy00** [reply]
如果可以自定义将军的移动位置，那就太好了。
比如攻击结束以后，将将军移动到自己想要的位置，位置是预先保存好的。

**Reply by skelgaard** [reply]
not possible... thats decided by the base game, so nothing that can be changed

**Comment by JohnRo79** [comment]
can you add a horseshoe timed option too pls?

Ex: after queuing up all generals, have a choice to apply shoes with timing too

**Reply by adly98** [reply]
well i think Sirris consider this a type of automation so it won't be allowed

**Reply by Mazimus332** [reply]
Why automation? It's just math automation, it's not game automation technically... Maybe auto-apply horseshoe would be automation, and I don't find it necessary. But selecting Horseshoes to change delay times depending on +% speed to make them fit better, this would make template exports much better and I support it.

**Reply by adly98** [reply]
ah, my bad i misunderstood what he was saying, well it is pretty much applicable

**Reply by Mazimus332** [reply]
Or maybe you're right, idk if I interpreted correctly, my contribution is there however :)

**Comment by Anprivate** [comment]
А можно как-то сделать, чтобы загрузка генералов была доступна и в том случае, когда не все генералы доступны? В таком случае загружать только тех генералов, которые доступны (даже если они стоят не на своих местах). Сильно ускорило бы процессы во время евента.
И еще одна проблема - если часть генералов в пресете в данный момент недоступна (перемещаются, в атаке, еще не прибыли), то загружаемые в них войска всё равно учитываются в необходимых. А это неправильно, так как считаться должны только те, которые доступны к загрузке в данный момент.
И еще - очень бы хотелось, чтобы кнопочка атаки была недоступна, если какие-либо генералы из назначенных к атаке были бы недоступны/неправильно загружены/не на своём месте.

UPD: что-то пока вообще не могу понять алгоритм появления кнопок MOVE/LOAD/ATTACK. В некоторых случаях вроде всё красное и все кнопки есть. В некоторых случаях вроде всё ок - кнопок нет. Ничего не понимаю. Попробую скриншотов наделать характерных
